include/albatross/src/graph/minimum_spanning_tree.hpp

```diff
@@ -173,6 +173,92 @@ Graph<VertexType> maximum_spanning_tree(const Graph<VertexType> &graph) {
   return output;
 }
 
+// A helper struct to track if a closed loop would be formed by a proposed edge
+// during Kruskal's algorithm. This struct associates each node with a tree,
+// initialized to each node in a unique tree. When a new edge is added, the
+// trees associated with the edge's nodes are merged into one tree. If a
+// proposed edge already has both nodes in the same tree, a closed loop would be
+// formed, and the edge can be rejected.
+template <typename VertexType> struct KruskalNodeTracker {
+
+  struct VertexWithTreeID {
+    VertexWithTreeID(const VertexType &v_, const size_t &tree_)
+        : v(v_), tree(tree_){};
+    VertexType v;
+    size_t tree;
+  };
+
+  KruskalNodeTracker(const std::set<VertexType> &vertices) : vertices_() {
+    size_t tree_id = 0;
+    for (const auto &v : vertices) {
+      vertices_.emplace_back(v, tree_id);
+      tree_id++;
+    }
+  };
+
+  bool in_same_tree_else_add_edge(const VertexType &u, const VertexType &v) {
```

Comment:
What does the `else_add_edge` part of this mean?

Reply:
There's two main steps of the algorithm.
(1) Check if a proposed edge forms a closed loop.
   - Yes - ignore the edge.
   - No add edge to MST and (2) merge the associated trees.
  Since (1) and (2) have repeated work, I would like to do them in one function call. But I can modify things to make this call more intuitive.

Reply:
I think the part that confused me is that it says `else_add_edge` but this function doesn't add an edge, that part happens outside this function, right? My understanding of this function is closer to: `is_same_tree_update_if_not`, but yeah I guess you could also split it into parts, things like:
```
const auto pair = find_vertex_and_tree(e.a, e.b);
if (!is_same_tree(pair.first, pair.second)) {
  update_trees(pair.first, pair.second);
  add_edge(e);
}
```